glib/src/object.rs

```diff
@@ -3267,6 +3279,96 @@ wrapper! {
     }
 }
 
+// rustdoc-stripper-ignore-next
+/// A weak reference to an object.
+///
+/// The difference to `WeakRef` is that it isn't thread safe and that it allows to connect a
+/// closure that is called when the set object is disposed.
+#[derive(Debug)]
+pub struct WeakRefLocal<T: ObjectType, F: Fn() + 'static>(
+    *mut (*mut GObject, F),
+    gobject_ffi::GWeakNotify,
+    PhantomData<*mut T>,
+);
+
+impl<T: ObjectType, F: Fn() + 'static> WeakRefLocal<T, F> {
+    /// Create a new empty weak reference.
+    ///
+    /// `upgrade` will always return `None` until an object is set on it.
+    /// The closure will be called when the set object is disposed.
+    pub fn new(f: F) -> WeakRefLocal<T, F> {
+        unsafe extern "C" fn notify_func<F: Fn()>(
+            data: ffi::gpointer,
+            _obj: *mut gobject_ffi::GObject,
+        ) {
+            let data = (data as *mut (*mut GObject, F)).as_mut().unwrap();
+
+            data.0 = ptr::null_mut();
+
+            let f: &F = &*(&data.1 as *const F);
+
+            f();
+        }
+
+        WeakRefLocal(
+            Box::into_raw(Box::new((ptr::null_mut(), f))),
+            Some(notify_func::<F>),
+            PhantomData,
+        )
+    }
+
+    // rustdoc-stripper-ignore-next
+    /// Set this weak reference to the given object.
+    #[doc(alias = "g_weak_ref_set")]
+    pub fn set(&self, obj: Option<&T>) {
+        unsafe {
+            let data = self.0.as_mut().unwrap();
+
+            if !data.0.is_null() {
+                gobject_ffi::g_object_weak_unref(data.0, self.1, self.0 as *mut _);
+            }
+
+            if let Some(obj) = obj {
+                let obj_ptr = obj.as_object_ref().to_glib_none().0;
+                data.0 = obj_ptr;
+                gobject_ffi::g_object_weak_ref(data.0, self.1, self.0 as *mut _);
+            } else {
+                data.0 = ptr::null_mut();
+            }
+        }
+    }
+
+    // rustdoc-stripper-ignore-next
+    /// Try to upgrade this weak reference to a strong reference.
+    ///
+    /// If the stored object was already destroyed or no object was set in this weak reference then
+    /// `None` is returned.
+    pub fn upgrade(&self) -> Option<T> {
```

Comment:
This is unsound and you really want to use `WeakRef` for this case. I wouldn't provide this feature here :)

Reply:
dropped it as well.